### PR TITLE
Bkarastanev/fix GitHub issue

### PR DESCRIPTION
**Pull request by borisKarastanev**
- Revert getTemplates() to use metadata from github

- Added tests for Template backup data

- Added tests for Page backup data

### CODE REVIEW

File: lib/services/template-service.ts [hunk 1]
```diff
@@ -83,97 +82,89 @@ export class TemplateService implements ITemplateService {
...
-            //                 .catch((errorTemplates: any) => {
-            //                     console.error(errorTemplates);
-            //                 });
-            //         } else {
-            //             // Load data from cache
-            //             resolve(value);
-            //         }
-            //     }
-            // });
+            tmpCache.get("tempDetails", (error: any, value: any) => {
```
Comment (manoldonev) [inline lib/services/template-service.ts]: `tmpCache` and `templDetails` abbreviations are unnecessarily confusing -- it is a temp[orary] cache or a template cache, temp[orary] details or template details?

Comment (borisKarastanev) [reply lib/services/template-service.ts]: Agreed

File: lib/services/template-service.ts [hunk 2]
```diff
@@ -83,97 +82,89 @@ export class TemplateService implements ITemplateService {
...
+                if (!error) {
```
Comment (manoldonev) [inline lib/services/template-service.ts]: I think we should consider refactoring / restructuring / reordering this method as its indentation is on its way to the pyramid of doom.

File: lib/services/template-service.ts [hunk 3]
```diff
@@ -83,97 +82,89 @@ export class TemplateService implements ITemplateService {
...
+                        // Load data from cache
+                        resolve(value);
+                    }
+                }
+            });
+        });
+    }
+
+    private sortTmpData(templates: Array<any>) {
+        const flavOrder: Array<string> = ["JavaScript", "TypeScript", "Angular & TypeScript"];
```
Comment (manoldonev) [inline lib/services/template-service.ts]: I think we should agree against using this kind of abbreviations -- in my opinion we are not gaining anything by calling this `flavOrder` vs `flavorOrder`

File: lib/services/template-service.ts [hunk 4]
```diff
@@ -83,97 +82,89 @@ export class TemplateService implements ITemplateService {
...
+            "Master-Detail with Kinvey"
+        ];
+        let sortedByType: Array<any>;
+        let sortByFlav: Array<any>;
+
+        sortedByType = util.sortBy(templates, (temp: any) => {
+            return util.indexOf(typeOrder, temp.displayName);
+        });
+
+        sortByFlav = util.sortBy(sortedByType, (temp: any) => {
```
Comment (manoldonev) [inline lib/services/template-service.ts]: Same comment about the `flav` abbreviation as above.

File: lib/services/template-service.ts (hunk 2 above)
Comment (borisKarastanev) [reply lib/services/template-service.ts]: I will think of a way to refactor this method, however, all those checks are necessary in order to provide the template data from Cache, Github, or fallback. Maybe we could think of something when we refactor the Caching logic in a separate PR?

File: lib/services/template-service.ts [hunk 5]
```diff
@@ -83,97 +82,89 @@ export class TemplateService implements ITemplateService {
...
-            //                 .catch((errorTemplates: any) => {
-            //                     console.error(errorTemplates);
-            //                 });
-            //         } else {
-            //             // Load data from cache
-            //             resolve(value);
-            //         }
-            //     }
-            // });
+            templateCache.get("templateDetails", (error: any, value: any) => {
```
Comment (manoldonev) [inline lib/services/template-service.ts]: Ideas for refactoring (as a separate PR):

- reverse the "master" if clause from:
``` ts
if (!error) {
...
} else {
   resolve(...)
}
```
to
``` ts
if (error) {
   resolve(...)
   return;
}

...
```
In this way you are decreasing the indentation with one step for the bulk of the method.

Also, we should consider whether it is better to write single-line thennables with the short form:
``` ts
then((details) => templateDetails.push(details)
```
vs
``` ts
then((details) => {
  templateDetails.push(details);
}
```

File: lib/services/template-service.ts [hunk 6]
```diff
@@ -83,97 +82,89 @@ export class TemplateService implements ITemplateService {
...
+                if (!error) {
+                    if (value === undefined) {
+                        this.getTemplatesNames()
+                            .then((templateNames: any) => {
```
Comment (manoldonev) [inline lib/services/template-service.ts]: This then statement does not do anything at the moment -- remove?

Comment (borisKarastanev) [reply lib/services/template-service.ts]: Removed

**Pull request merged**